bombadillo.1
.TH "bombadillo" 1 "27 OCT 2019" "" "General Operation Manual"
.SH NAME
\fBbombadillo \fP- a non-web browser
.SH SYNOPSIS
.nf
.fam C
\fBbombadillo\fP [\fIoptions\fP] [\fIurl\fP]
.fam T
.fi
.SH DESCRIPTION
\fBbombadillo\fP is a non-web browser for the terminal. It features a full terminal user interface, vim-like keybindings, document pager, configurable settings, and a robust command selection.
.TP
\fBbombadillo\fP supports the following protocols as first class citizens: gopher, gemini, finger, and local (a user’s file system). Support for telnet, http and https is also available via integration with third party applications.
.SH OPTIONS
.TP
.B
\fB-h\fP
Display usage help and exit. Provides a list of all command line options with a short description and exits.
.TP
.B
\fB-t\fP
Set the window title to 'Bombadillo'. Can be used in a GUI environment, however not all terminals support this feature.
.TP
.B
\fB-v\fP
Display version information and exit.
.SH PROTOCOL SUPPORT
All of the below protocols are supported. With the exception of gopher, the protocol name must be present as the scheme component of a url in the form of \fI[protocol]://[the rest of the url]\fP.
.TP
.B
gopher
Gopher is the default protocol for \fBbombadillo\fP. Any textual item types will be visited and shown to the user and any non-text types will be downloaded. Type 7 (querying) is fully supported. As the default protocol, any url that is not prefixed with the scheme section of a url (\fIgopher://\fP for example) will be treated as gopher urls.
.TP
.B
gemini
Gemini is supported, but as a new protocol with an incomplete specification, features may change over time. At present Bombadillo supports TLS with a trust on first use certificate pinning system (similar to SSH). Gemini maps and other text types are rendered in the browser and non-text types will be downloaded.
.TP
.B
finger
Basic support is provided for the finger protocol. The format is: \fIfinger://[[username@]][hostname]\fP. Many servers still support finger and it can be fun to see if friends are online or read about the users whose phlogs you follow.
.TP
.B
local
Local is similar to the \fIfile\fP protocol used in web browsers or the like, with a smaller set of features. Users can use the local scheme to view files on their local system. Directories are supported as viewable text object as well as any files. Wildcards and globbing are not supported. Using \fI~\fP to represent a user's home directory, as well as relative paths, are supported. The \fIcolor\fP theme has no effect on this protocol and all terminal escape sequences will be rendered to the screen literally.
.TP
.B
telnet
Telnet is not supported directly, but addresses will be followed and opened as a subprocess by whatever telnet client a user sets in their settings (defaulting to \fItelnet\fP). In some cases this behavior may be buggy.
.TP
.B
http, https
Neither of the world wide web protocols are supported directly. \fBbombadillo\fP can be configured to open web links in a user's default graphical web browser. It is also possible to display web content directly in \fBbombadillo\fP using lynx, w3m, or elinks terminal web browsers to render pages. Opening http/https links is opt-in only, controlled by the \fIwebmode\fP setting.
.IP
Opening links in a default graphical web browser will only work in a GUI environment.
.IP
Displaying web content directly in \fBbombadillo\fP requires lynx, w3m or elinks terminal web browsers are installed on the system.
.SH COMMANDS
.SS  KEY COMMANDS
These commands work as a single keypress anytime \fBbombadillo\fP is not taking in a line based command or when the user is being prompted for action. This is the default command mode of \fBbombadillo\fP.
.TP
.B
b, h
Navigate back one place in your document history.
.TP
.B
B
Toggle the bookmarks panel open/closed.
.TP
.B
d
Scroll down an amount corresponding to 75% of your terminal window height in the current document.
.TP
.B
f, l
Navigate forward one place in your document history.
.TP
.B
g
Scroll to the top of the current document.
.TP
.B
G
Scroll to the bottom of the current document.
.TP
.B
j
Scroll down a single line in the current document.
.TP
.B
k
Scroll up a single line.
.TP
.B
n
Jump to next found text item.
.TP
.B
N
Jump to previous found text item.
.TP
.B
q
Quit \fBbombadillo\fP.
.TP
.B
R
Reload the current page (does not destroy forward history).
.TP
.B
1, 2, 3, 4, 5, 6, 7, 8, 9, 0
Quick navigation to the first 10 links on a page. The 0 key will navigate to the link numbered '10', all other numbers navigate to their matching link number.
.TP
.B
u
Scroll up an amount corresponding to 75% of your terminal window height in the current document.
.TP
.B
/
Search for text within current document. / followed by a text query will highlight and allow navigation of found text. / with an empty query will clear the current query.
.TP
.B
<tab>
Toggle the scroll focus between the bookmarks panel and the document panel. Only has an effect if the bookmarks panel is open.
.TP
.B
<spc>
Enter line command mode. Once a line command is input, the mode will automatically revert to key command mode.
.TP
.B
:
Alias for <spc>. Enter line command mode.
.SS  LINE COMMANDS
These commands are typed in by the user to perform an action of some sort. As listed in KEY COMMANDS, this mode is initiated by pressing : or <space>. The command names themselves are not case sensitive, though the arguments supplied to them may be.
.TP
.B
[url]
Navigates to the requested url.
.TP
.B
[link id]
Follows a link on the current document with the given number.
.TP
.B
add [url] [name\.\.\.]
Adds the url as a bookmarks labeled by name. \fIa\fP can be used instead of the full \fIadd\fP.
.TP
.B
add [link id] [name\.\.\.]
Adds the url represented by the link id within the current document as a bookmark labeled by name. \fIa\fP can be used instead of the full \fIadd\fP.
.TP
.B
add . [name\.\.\.]
Adds the current document's url as a bookmark labeled by name. \fIa\fP can be used instead of the full \fIadd\fP.
.TP
.B
bookmarks
Toggles the bookmarks panel open/closed. Alias for KEY COMMAND \fIB\fP. \fIb\fP can be used instead of the full \fIbookmarks\fP.
.TP
.B
bookmarks [bookmark id]
Navigates to the url represented by the bookmark matching bookmark id. \fIb\fP can be entered, rather than the full \fIbookmarks\fP.
.TP
.B
check [link id]
Displays the url corresponding to a given link id for the current document. \fIc\fP can be used instead of the full \fIcheck\fP.
.TP
.B
check [setting name]
Displays the current value for a given configuration setting. \fIc\fP can be used instead of the full \fIcheck\fP.
.TP
.B
delete [bookmark id]
Deletes the bookmark matching the bookmark id. \fId\fP can be used instead of the full \fIdelete\fP.
.TP
.B
help
Navigates to the gopher based help page for \fBbombadillo\fP. \fI?\fP can be used instead of the full \fIhelp\fP.
.TP
.B
home
Navigates to the document set by the \fIhomeurl\fP setting. \fIh\fP can be entered, rather than the full \fIhome\fP.
.TP
.B
jump
Navigates to the previous page in history from the current page. Useful for keeping the current page in your history while still browsing. \fIj\fP can be used instead of the full \fIjump\fP.
.TP
.B
jump [history location]
Navigates to the given history location. The history location should be an integer between 0 and 20. \fIj\fP can be used instead of the full \fIjump\fP.
.TP
.B
purge *
Deletes all pinned gemini server certificates. \fIp\fP can be used instead of the full \fIpurge\fP.
.TP
.B
purge [host name]
Deletes the pinned gemini server certificate for the given hostname. \fIp\fP can be used instead of the full \fIpurge\fP.
.TP
.B
quit
Quits \fBbombadillo\fP. Alias for KEY COMMAND \fIq\fP. \fIq\fP can be used instead of the full \fIquit\fP.
.TP
.B
reload
Requests the current document from the server again. This does not break forward history the way entering the url again would. \fIr\fP can be used instead of the full \fIreload\fP.
.TP
.B
search
Queries the user for search terms and submits a search to the search engine set by the \fIsearchengine\fP setting.
.TP
.B
search [keywords\.\.\.]
Submits a search to the search engine set by the \fIsearchengine\fP setting, with the query being the provided keyword(s).
.TP
.B
set [setting name] [value]
Sets the value for a given configuration setting. \fIs\fP can be used instead of the full \fIset\fP.
.TP
.B
version
Shows the current Bombadillo version number.
.TP
.B
write .
Writes the current document to a file. The file is named by the last component of the url path. If the last component is blank or \fI/\fP a default name will be used. The file saves to the directory set by the \fIsavelocation\fP setting. \fIw\fP can be entered rather than the full \fIwrite\fP.
.TP
.B
write [url]
Writes data from a given url to a file. The file is named by the last component of the url path. If the last component is blank or \fI/\fP a default name will be used. The file saves to the directory set by the \fIsavelocation\fP setting. \fIw\fP can be entered rather than the full \fIwrite\fP.
.TP
.B
write [link id]
Writes data from a given link id in the current document to a file. The file is named by the last component of the url path. If the last component is blank or \fI/\fP a default name will be used. The file saves to the directory set by the \fIsavelocation\fP setting. \fIw\fP can be entered rather than the full \fIwrite\fP.
.SH FILES
\fBbombadillo\fP keeps a hidden configuration file in a user's XDG configuration directory. The file is a simplified ini file titled \fI.bombadillo.ini\fP. It is generated when a user first loads \fBbombadillo\fP and is updated with bookmarks and settings as a user adds them. The file can be directly edited, but it is best to use the SET command to update settings whenever possible. To return to the state of a fresh install, simply remove the file and a new one will be generated with the \fBbombadillo\fP defaults. On some systems an administrator may set the configuration file location to somewhere other than the default setting. If you do not see the file where you expect it, or if your settings are not being read, try \fI:check configlocation\fP to see where the file should be, or contact your system administrator for more information.
.SH SETTINGS
The following is a list of the settings that \fBbombadillo\fP recognizes, as well as a description of their valid values.
.TP
.B
configlocation
The path to the directory that the \fI.bombadillo.ini\fP configuration file is stored in. This is a \fBread only\fP setting and cannot be changed with the \fIset\fP command, but it can be read with the \fIcheck\fP command.
.TP
.B
defaultscheme
The scheme that should be used when no scheme is present in a given URL. \fIgopher\fP, \fIgemini\fP, \fIhttp\fP, and \fIhttps\fP are valid values.
.TP
.B
geminiblocks
Determines how to treat preformatted text blocks in text/gemini documents. \fIblock\fP will show the contents of the block, \fIalt\fP will show any available alt text for the block, \fIboth\fP will show both the content and the alt text, and \fIneither\fP will show neither. Unlike other settings, a change to this value will require a fresh page load to see the change.
.TP
.B
homeurl
The url that \fBbombadillo\fP navigates to when the program loads or when the \fIhome\fP or \fIh\fP LINE COMMAND is issued. This should be a valid url. If a scheme/protocol is not included, gopher will be assumed.
.TP
.B
savelocation
The path to the directory that \fBbombadillo\fP should write files to. This must be a valid filepath for the system, must be a directory, and must already exist.
.TP
.B
searchengine
The url to use for the LINE COMMAND \fIsearch\fP. Should be a valid search path that terms may be appended to.
.TP
.B
telnetcommand
Tells the browser what command to use to start a telnet session. Should be a valid command, including any flags. The address being navigated to will be added to the end of the command.
.TP
.B
theme
Can toggle between visual modes. Valid values are \fInormal\fP, \fIcolor\fP, and \fIinverse\fP. When set to inverse, the normal mode colors are inverted. Both normal and inverse modes filter out terminal escape sequences. When set to color, Bombadillo will render terminal escape sequences representing colors when it finds them in documents.
.TP
.B
timeout
The number of seconds after which connections to gopher or gemini servers should time out if the server has not responded.
.TP
.B
webmode
Controls behavior when following web links. The following values are valid: \fInone\fP will disable following web links, \fIgui\fP will have the browser attempt to open web links in a user's default graphical web browser; \fIlynx\fP, \fIw3m\fP, and \fIelinks\fP will have the browser attempt to use the selected terminal web browser to handle the rendering of web pages and will display the pages directly in Bombadillo.

.SH BUGS
There are very likely bugs. Many known bugs can be found in the issues section of \fBbombadillo\fP's source code repository (see \fIlinks\fP).
.SH LINKS
\fBbombadillo\fP maintains a presence in the following locations:
.TP
.B
Source Code Repository
https://tildegit.org/sloum/bombadillo
.TP
.B
Web Homepage
http://bombadillo.colorfield.space
.TP
.B
Gopher Homepage
gopher://bombadillo.colorfield.space
.SH AUTHORS
\fBbombadillo\fP was primarily developed by sloum, with kind and patient assistance from ~asdf and jboverf.


bombadillo.go
package main

import (
	"bombadillo/gopher"
	"bombadillo/cmdparse"
	"bombadillo/config"
	"bombadillo/cui"
	"os/user"
	"io/ioutil"
	"os"
	"fmt"
	"strings"
	"regexp"
	"strconv"
)

var helplocation string ="gopher://colorfield.space:70/1/bombadillo-info"
var history gopher.History = gopher.MakeHistory()
var screen *cui.Screen
var userinfo, _ = user.Current()
var settings config.Config
var options = map[string]string{
	"homeurl": "gopher://colorfield.space:70/1/bombadillo-info",
	"savelocation": userinfo.HomeDir + "/Downloads/",
	"searchengine": "gopher://gopher.floodgap.com:70/7/v2/vs",
	"openhttp": "false",
	"httpbrowser": "lynx",
}

func err_exit(err string, code int) {
	fmt.Println(err)
	os.Exit(code)
}

func save_file(address, name string) error {
	quickMessage("Saving file...", false)
	defer quickMessage("Saving file...", true)

	url, err := gopher.MakeUrl(address)
	if err != nil {
		return err
	}

	data, err := gopher.Retrieve(url)
	if err != nil {
		return err
	}

	err = ioutil.WriteFile(options["savelocation"] + name, data, 0644)
	if err != nil {
		return err
	}

	return fmt.Errorf("Saved file to " + options["savelocation"] + name)
}

func save_file_from_data(v gopher.View) error {
	urlsplit := strings.Split(v.Address.Full, "/")
	filename := urlsplit[len(urlsplit) - 1]
	save_msg := fmt.Sprintf("Saved file as %q", options["savelocation"] + filename)
	quickMessage(save_msg, false)
	defer quickMessage(save_msg, true)
	err := ioutil.WriteFile(options["savelocation"] + filename, []byte(strings.Join(v.Content, "")), 0644)
	if err != nil {
		return err
	}

	return fmt.Errorf(save_msg)
}

func search(u string) error {
	cui.MoveCursorTo(screen.Height - 1, 0)
	cui.Clear("line")
	fmt.Print("Enter form input: ")
	cui.MoveCursorTo(screen.Height - 1, 17)
	entry := cui.GetLine()
	quickMessage("Searching...", false)
	searchurl := fmt.Sprintf("%s\t%s", u, entry)
	sv, err := gopher.Visit(searchurl, options["openhttp"])
	if err != nil {
		quickMessage("Searching...", true)
		return err
	}
	history.Add(sv)
	quickMessage("Searching...", true)
	updateMainContent()
	screen.Windows[0].Scrollposition = 0
	screen.ReflashScreen(true)
	return nil
}


func route_input(com *cmdparse.Command) error {
	var err error
	switch com.Type {
		case cmdparse.SIMPLE:
			err = simple_command(com.Action)
		case cmdparse.GOURL:
			err = go_to_url(com.Target)
		case cmdparse.GOLINK:
			err = go_to_link(com.Target)
		case cmdparse.DOLINK:
			err = do_link_command(com.Action, com.Target)
		case cmdparse.DOAS:
			err = do_command_as(com.Action, com.Value)
		case cmdparse.DOLINKAS:
			err = do_link_command_as(com.Action, com.Target, com.Value)
		default:
			return fmt.Errorf("Unknown command entry!")
	}

	return err
}

func toggle_bookmarks() {
	bookmarks := screen.Windows[1]
	main := screen.Windows[0]
	if bookmarks.Show {
		bookmarks.Show = false
		screen.Activewindow = 0
		main.Active = true
		bookmarks.Active = false
	} else {
		bookmarks.Show = true
		screen.Activewindow = 1
		main.Active = false
		bookmarks.Active = true
	}

	if screen.Activewindow == 0 {
	} else {
	}
	screen.ReflashScreen(false)
}

func simple_command(a string) error {
	a = strings.ToUpper(a)
	switch a {
		case "Q", "QUIT":
			cui.Exit()
		case "H", "HOME":
			return go_home()
		case "B", "BOOKMARKS":
			toggle_bookmarks()
		case "SEARCH":
			return search(options["searchengine"])
		case "HELP":
			return go_to_url(helplocation)

		default:
			return fmt.Errorf("Unknown action %q", a)
	}
	return nil
}

func go_to_url(u string) error {
	quickMessage("Loading...", false)
	v, err := gopher.Visit(u, options["openhttp"])
	if err != nil {
		quickMessage("Loading...", true)
		return err
	}
	quickMessage("Loading...", true)

	if v.Address.Gophertype == "7" {
		err := search(v.Address.Full)
		if err != nil {
			return err
		}
	} else if v.Address.IsBinary {
		// TO DO: run this into the write to file method
		return save_file_from_data(v)
	} else {
		history.Add(v)
	}
	updateMainContent()
	screen.Windows[0].Scrollposition = 0
	screen.ReflashScreen(true)
	return nil
}

func go_to_link(l string) error {
	if num, _ := regexp.MatchString(`^\d+$`, l); num && history.Length > 0 {
		linkcount := len(history.Collection[history.Position].Links)
		item, _ := strconv.Atoi(l)
		if item <= linkcount {
			linkurl := history.Collection[history.Position].Links[item - 1]
			quickMessage("Loading...", false)
			v, err := gopher.Visit(linkurl, options["openhttp"])
			if err != nil {
				quickMessage("Loading...", true)
				return err
			}
			quickMessage("Loading...", true)

			if v.Address.Gophertype == "7" {
				err := search(linkurl)
				if err != nil {
					return err
				}
			} else if v.Address.IsBinary {
				return save_file_from_data(v)
			} else {
				history.Add(v)
			}
		} else {
			return fmt.Errorf("Invalid link id: %s", l)
		}
	} else {
			return fmt.Errorf("Invalid link id: %s", l)
	}
	updateMainContent()
	screen.Windows[0].Scrollposition = 0
	screen.ReflashScreen(true)
	return nil
}

func go_home() error {
	if options["homeurl"] != "unset" {
		return go_to_url(options["homeurl"])
	}
	return fmt.Errorf("No home address has been set")
}

func do_link_command(action, target string) error {
	num, err := strconv.Atoi(target)
	if err != nil {
		return fmt.Errorf("Expected number, got %q", target)
	}

	switch action {
		case "DELETE", "D":
			err := settings.Bookmarks.Del(num)
			screen.Windows[1].Content = settings.Bookmarks.List()
			save_config()
			screen.ReflashScreen(false)
			return err
		case "BOOKMARKS", "B":
			if num > len(settings.Bookmarks.Links) - 1 {
				return fmt.Errorf("There is no bookmark with ID %d", num)
			}
			err := go_to_url(settings.Bookmarks.Links[num])
			return err
	}

	return fmt.Errorf("This method has not been built")
}

func do_command_as(action string, values []string) error {
	if len(values) < 2 {
		return fmt.Errorf("%q", values)
	}

	if values[0] == "." {
		values[0] = history.Collection[history.Position].Address.Full
	}

	switch action {
		case "ADD", "A":
			err := settings.Bookmarks.Add(values)
			if err != nil {
				return err
			}
			screen.Windows[1].Content = settings.Bookmarks.List()
			save_config()
			screen.ReflashScreen(false)
			return nil
		case "WRITE", "W":
			return  save_file(values[0], strings.Join(values[1:], " "))
		case "SET", "S":
			if _, ok := options[values[0]]; ok {
				options[values[0]] = strings.Join(values[1:], " ")
				save_config()
				return nil
			}
			return fmt.Errorf("Unable to set %s, it does not exist",values[0])
	}
	return fmt.Errorf("Unknown command structure")
}

func do_link_command_as(action, target string, values []string) error {
	num, err := strconv.Atoi(target)
	if err != nil {
		return fmt.Errorf("Expected number, got %q", target)
	}

	links := history.Collection[history.Position].Links
	if num >= len(links) {
		return fmt.Errorf("Invalid link id: %s", target)
	}

	switch action {
		case "ADD", "A":
			newBookmark := append([]string{links[num - 1]}, values...)
			err := settings.Bookmarks.Add(newBookmark)
			if err != nil {
				return err
			}
			screen.Windows[1].Content = settings.Bookmarks.List()
			save_config()
			screen.ReflashScreen(false)
			return nil
		case "WRITE", "W":
			return save_file(links[num - 1], strings.Join(values, " "))
	}

	return fmt.Errorf("This method has not been built")
}


func updateMainContent() {
	screen.Windows[0].Content = history.Collection[history.Position].Content
	screen.Bars[0].SetMessage(history.Collection[history.Position].Address.Full)
}

func clearInput(incError bool) {
	cui.MoveCursorTo(screen.Height - 1, 0)
	cui.Clear("line")
	if incError {
		cui.MoveCursorTo(screen.Height, 0)
		cui.Clear("line")
	}
}

func quickMessage(msg string, clearMsg bool) {
	cui.MoveCursorTo(screen.Height, screen.Width - 2 - len(msg))
	if clearMsg {
		cui.Clear("right")
	} else {
		fmt.Print("\033[48;5;21m\033[38;5;15m", msg, "\033[0m")
	}
}


func save_config() {
	bkmrks := settings.Bookmarks.IniDump()
	opts := "\n[SETTINGS]\n"
	for k, v := range options {
		opts += k
		opts += "="
		opts += v
		opts += "\n"
	}
	ioutil.WriteFile(userinfo.HomeDir + "/.bombadillo.ini", []byte(bkmrks+opts), 0644)
}


func load_config() {
	file, err := os.Open(userinfo.HomeDir + "/.bombadillo.ini")
	if err != nil {
		save_config()
	}
	confparser := config.NewParser(file)
	settings, _ = confparser.Parse()
	file.Close()
	screen.Windows[1].Content = settings.Bookmarks.List()
	for _, v := range settings.Settings {
		lowerkey := strings.ToLower(v.Key)
		if _, ok := options[lowerkey]; ok {
			options[lowerkey] = v.Value
		}
	}
}

func toggleActiveWindow(){
	if screen.Windows[1].Show {
		if screen.Windows[0].Active {
			screen.Windows[0].Active = false
			screen.Windows[1].Active = true
			screen.Activewindow = 1
		} else {
			screen.Windows[0].Active = true
			screen.Windows[1].Active = false
			screen.Activewindow = 0
		}
		screen.Windows[1].DrawWindow()
	}
}


func display_error(err error) {
	cui.MoveCursorTo(screen.Height, 0)
	fmt.Print("\033[41m\033[37m", err, "\033[0m")
}


func initClient() {
	history.Position = -1
	screen = cui.NewScreen()
	cui.SetCharMode()
	screen.AddWindow(2, 1, screen.Height - 2, screen.Width, false, false, true)
	screen.Windows[0].Active = true
	screen.AddMsgBar(1, "  ((( Bombadillo )))  ", "  A fun gopher client!", true)
	bookmarksWidth := 40
	if screen.Width < 40 {
		bookmarksWidth = screen.Width
	}
	screen.AddWindow(2, screen.Width - bookmarksWidth, screen.Height - 2, screen.Width, false, true, false)
	load_config()
}


func main() {
	defer cui.Exit()
	initClient()
	mainWindow := screen.Windows[0]
	first_load := true

	for {
		if first_load {
			first_load = false
			err := go_home()

			if err == nil {
				updateMainContent()
			}
			continue
		}

		c := cui.Getch()
		switch c {
			case 'j', 'J':
				screen.Windows[screen.Activewindow].ScrollDown()
				screen.ReflashScreen(false)
			case 'k', 'K':
				screen.Windows[screen.Activewindow].ScrollUp()
				screen.ReflashScreen(false)
			case 'q', 'Q':
				cui.Exit()
			case 'b':
				success := history.GoBack()
				if success {
					mainWindow.Scrollposition = 0
					updateMainContent()
					screen.ReflashScreen(true)
				}
			case 'B':
				toggle_bookmarks()
			case 'f', 'F':
				success := history.GoForward()
				if success {
					mainWindow.Scrollposition = 0
					updateMainContent()
					screen.ReflashScreen(true)
				}
			case '\t':
				toggleActiveWindow()
			case ':',' ':
				cui.MoveCursorTo(screen.Height - 1, 0)
				entry := cui.GetLine()
				// Clear entry line and error line
				clearInput(true)
				if entry == "" {
					continue
				}
				parser := cmdparse.NewParser(strings.NewReader(entry))
				p, err := parser.Parse()
				if err != nil {
					display_error(err)
				} else {
					err := route_input(p)
					if err != nil {
						display_error(err)
					}
				}
		}
	}
}
